### ISSUE TITLE
Xpand2! and Realearn

### ISSUE THREAD

**Issue by Realearnxpander** [helgoboss/realearn#790]
Hello community,

I'm trying to switch presets with Realearn, in the VST Xpand2! It works perfectly but sometimes all my notes sustain. I don't know what causes this but this way i can't use this in live situations... i hope theres a solution

**Comment by helgoboss** [comment]
Do you use the "FX: Browse presets" target? Anyway, I guess this is more related to Xpand2 than to ReaLearn?

**Comment by Realearnxpander** [comment]
Yes exactly that target i use, is that correct? I don't know if it is related to Xpand2! but I wish i knew the fix for it.

**Comment by Realearnxpander** [comment]
it occurs sometimes when I use my pads on the midicontroller to switch to other presets. And to stop the sustaining i have to play the note(s) that are sustained. After that it stops.

**Comment by Realearnxpander** [comment]
And I set the parameters in target min/max to the same value to load the right preset. I don't know any other way to do it.

**Comment by helgoboss** [comment]
I think the sustain issue is Xpand2's fault. But you could trigger REAPER action "Send all-notes-off and all-sounds-off to all MIDI outputs/plug-ins" (using another mapping with the same button as source and target "Project: Invoke REAPER action"). That would reset all notes before switching the preset ... big hammer, but could work.

Actually, "FX: Browse presets" is not really suitable if you want to recall a specific preset, it's more intended for browsing in the list of presets (with prev/next buttons or encoders). Setting target min/max easily breaks if you add/remove presets. If I want to map presets to particular buttons, I usually use "FX: Load snapshot". Pressing "Take!" takes the snapshot. You can then load it by triggering the target with the button. What's cool about it is that it doesn't depend at all on the preset list. The preset is saved as part of your project.

**Comment by Realearnxpander** [comment]
Thanks alot for your valuable information. Is there a tutorial on het FX: Load snapshot"? Because I have no clue what you mean, but it means there is a better way!

**Comment by Realearnxpander** [comment]
I have read about the 'Send all-notes-off and all-sounds-off to all MIDI outputs/plug-ins' I will look into that. I get the idea with the mapping, you can invoke that action to every button on which load a preset. It sounds like it could work. Brilliant! How how do i get that specific action into a mapping? I have to go to the actionlist for that normally

**Comment by helgoboss** [comment]
> Thanks alot for your valuable information. Is there a tutorial on het FX: Load snapshot"? Because I have no clue what you mean, but it means there is a better way!

No, I thought it's kind of self-explaining together with the user guide: Choose target "FX: Load snapshot", choose the FX that you want to take/load snapshots for, press "Take!" once to memorize the preset and then it should recall that preset whenever you press the button which you define as source for that mapping.



> I have read about the 'Send all-notes-off and all-sounds-off to all MIDI outputs/plug-ins' I will look into that. I get the idea with the mapping, you can invoke that action to every button on which load a preset. It sounds like it could work. Brilliant! How how do i get that specific action into a mapping? I have to go to the actionlist for that normally

Target "Project: Invoke REAPER action", choose "Pick!" and yes, pick that action.

**Comment by Realearnxpander** [comment]
Thanks with the user guide and your explaination above I now now how to do both!! Thanks. I hope the send all notes off action works. havent had the problem yet, so far so good. 

Now I've got another question. For instance, i press button 1 and have 2/3 tracks on, but somewhere in the song I want another FX activated, when i change a preset by pressing button 2. Thus i have 3/3 tracks on. After changing preset by pressing button 3, i want track 3 to be muted automatically. Is there a way to add an action (only IF mapping 1 is activated, unmute track 3)?

**Comment by Realearnxpander** [comment]
You have been amazingly helpfull by the way, thanks so much

**Comment by helgoboss** [comment]
One way is to simply add another mapping to the same button that mutes track 3 (target "Track: (Un)Mute".

Another more sophisticated way is to use mapping snapshots. It works a bit like SWS snapshots but has the benefit of being integrated into ReaLearn. With that, you can recall the target values of multiple mappings at once at the press of a button. E.g. recall the arm states of your track. It's perfect for loading different setups for different songs. If you want that, please have a look into the user guide.

**Comment by Realearnxpander** [comment]
Thanks for your quick Reasponse (see what I did there)... I will look into that.

One more question about send all notes off. What type of mode do I need to choose for that action to work correctly with a pad? Does it need to be a button, or normal mode?

**Comment by Realearnxpander** [comment]
because the problem isn't solved yet
